custom-local-authenticator/src/main/java/org/wso2/custom/authenticator/local/CustomLocalAuthenticator.java

```diff
@@ -0,0 +1,162 @@
+                                                 AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        log.info("This is the initiateAuthenticationRequest method");
+        String loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL();
+        // This is the default WSO2 IS login page. If you can create your custom login page you can use
+        // that instead.
+        String queryParams =
+                FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                        context.getCallerSessionKey(),
+                        context.getContextIdentifier());
+        try {
+            String retryParam = "";
+
+            if (context.isRetrying()) {
+                retryParam = "&authFailure=true&authFailureMsg=login.fail.message";
+            }
+
+            response.sendRedirect(response.encodeRedirectURL(loginPage + ("?" + queryParams)) +
+                    "&authenticators=BasicAuthenticator:" + "LOCAL" + retryParam);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException(e.getMessage(), e);
+        }
+    }
+
+    /**
+     * This method is used to process the authentication response.
+     */
+    @Override
+    protected void processAuthenticationResponse(HttpServletRequest request,
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        log.info("This is the processAuthenticationResponse method");
+        String username = request.getParameter(USER_NAME);
+        boolean isAuthenticated = true;
+
+        if(isAuthenticated) {
+            boolean authorization = false;
+            if (OIDC.equalsIgnoreCase(context.getRequestType())) {
+                log.info("Authentication request type is OIDC. Engaging custom local authenticator...");
+                // authorization only for openid connect requests
+                try {
+                    int tenantId = CustomLocalAuthenticatorServiceComponent.getRealmService().getTenantManager().
+                            getTenantId(MultitenantUtils.getTenantDomain(username));
+                    UserStoreManager userStoreManager =
+                            (UserStoreManager) CustomLocalAuthenticatorServiceComponent.
+                                    getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+
+                    // verify user is assigned to role
+                    authorization = ((AbstractUserStoreManager) userStoreManager).isUserInRole(
+                            username, REQUIRED_ROLE);
+                } catch (UserStoreException e) {
+                    log.error(e);
+                }
+            }
+
+            if (!authorization) {
+                log.error("Authentication failed! User: " + username + " doesn't have the role: " + REQUIRED_ROLE);
+
```

Comment:
redundant new line. Please look at other places as well. Might be better to reformat the code as well